### ISSUE TITLE
Sort product list by 'on hand' amount

### ISSUE THREAD

**Issue by tschumilas** [openfoodfoundation/wishlist#149]
## What is the need / problem?
At present a user ca filter their product list (/admin/products) by producer, category, or import date.  And they can search for specific product.  And they can sort the product list by product name (alphabetically or reverse alphabetically).

For producers with large product lists (many on OFN-CAN have over 300 products for example),  none of these searches/filters are particularly helpful.   If a producer wants to know what products they are currenly selling via the hubs they supply for example,  or to see what stock they have remaining for example,  the producer needs to scroll through their list to see where on hand = greater than 0.   

A related problem is that when a producer can't deliver an ordered product to the hub (was of poor quality....),  the hub removes it from the customer's order,   and that unavailable product goes right back in the producer's on-hand list -  and is now available for sale again.    its easy for the producer with a large product list to miss these products that the hub returned to their product list -  and this results in products that are not avaialble being re-sold.   

The producer needs an easier way to see what products they have available for sale -  in all shops -  at any point in time.

Some of this will be helped by the planed product uplift  maybe ??  -- but in the meantime -  I'm really hoping it would be an easy peasy papercut? 

## Which type of users does this problem affect (and how many, if known)?
Producers -  esp those with large product lists

## Potential solutions that will solve the problem  ?[[brainstorming to list feature candidates]

Make it possible to order the 'on-hand' amount column on /admin/products -  in the same way that product name can be ordered -  so that at least all the products with '0' on hand will cluster,  and the producer can ignore and more easily find the products that are available for sale because on-hand = greater than 0.

## Is Papercut or feature candidate?

_Add papercut or feature label to answer the question. If you are not sure, don't worry product team will cover :-)_

**Comment by Matt-Yorkley** [comment]
I think the third paragraph is a totally separate issue to ordering products by on_hand;

> A related problem is that when a producer can't deliver an ordered product to the hub (was of poor quality....), the hub removes it from the customer's order, and that unavailable product goes right back in the producer's on-hand list - and is now available for sale again. its easy for the producer with a large product list to miss these products that the hub returned to their product list - and this results in products that are not available being re-sold.

We could potentially do something like: when a user is cancelling an order we could show a checkbox (checked by default) with a message like "Restock Items: return all items from this order to stock". That way it's clear what's happening, and there's an option to not restock the items, if that's not what the user wants to do.

**Comment by tschumilas** [comment]
Thanks Matt -  I've created a second issue for the returning unavailable items to stock issue.  https://github.com/openfoodfoundation/wishlist/issues/150

**Comment by RachL** [comment]
Papercut meeting note: 

> The producer needs an easier way to see what products they have available for sale - in all shops - at any point in time.

Sorting the column might not really easily fix that need, but it could be a good place to start.

- we need to be careful about on hand value when on demand is ticked. Infinity is a bigger number than 7, so order should be from 0 to infinite.

- how to handle variants? When one variant has less than another, what should be the rule to order products? 

@tschumilas did you have something in mind for variant vs product ordering?

**Comment by tschumilas** [comment]
I was thinking that clicing on 'on hand'  to order the products & variants would be treated the same as 'name' is now.  When I sort on 'name' -  it is by the product name,  and then the variants come with the product.      So,  if I have product:  carrots,  and there are 2 variants:  'red' -  on-hand = 0,   and  'orange' - on hand = 10,    and I order the on-hand amounts by clicking on the column label,  the product 'carrots'    I want to see that I have carrots available for sale --- so I think I want to see all the carrot variants.    I don't think it would make sense to split up the product  -  know what I mean?    (But I confess I haven't thought of this yet.....)

**Comment by RachL** [comment]
@tschumilas just to be sure I understand is this correct : 

- Product carrots has variant A with a stock of 0 and variant B with a stock of 10
- Product Banana has variant A with a stock of 4 and variant B with a stock of 5

With asc. order Banana would be **before** carrots.

**Comment by tschumilas** [comment]
yes thats what I would say.   happy to hear other ideas on this though....

**Comment by Matt-Yorkley** [comment]
Dev notes: this ordering is more tricky than it appears, due to the fact that the on_hand value for a product is a calculated value (not a fixed value on the object). It also involves joins and sums across multiple tables (variants and stock_items) which aren't part of how the data is currently queried.

**Comment by tschumilas** [comment]
Oh dear -  it seemed so simple initially.  Are there other ways a manager can find out how much stock of a product they have remaining,  at any point in time.   It honestly seems like such a basic need.......  We did a zap for one farm to help with this -  but it was much too costly....

**Comment by RachL** [comment]
Papercut meeting 02/02/2022: unfortunatiely this is too tricky to make it a papercut.

**Comment by RachL** [comment]
closing in favor of https://github.com/openfoodfoundation/openfoodnetwork/issues/13435
